src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -375,32 +373,40 @@ bool ObjectMonitor::enter_for(JavaThread* locking_thread) {
     return false;
   }
 
-  enter_for_with_contention_mark(locking_thread, contention_mark);
+  bool success = TryLockWithContentionMark(locking_thread, contention_mark);
+
+  assert(success, "Failed to enter_for: locking_thread=" INTPTR_FORMAT
+         ", this=" INTPTR_FORMAT "{owner=" INTPTR_FORMAT "}",
+         p2i(locking_thread), p2i(this), p2i(owner_raw()));
   assert(owner_raw() == locking_thread, "must be");
   return true;
 }
 
-bool ObjectMonitor::try_enter(JavaThread* current) {
-  // TryLock avoids the CAS
+bool ObjectMonitor::try_enter(JavaThread* current, bool check_owner) {
+  // TryLock avoids the CAS and handles deflation.
   TryLockResult r = TryLock(current);
   if (r == TryLockResult::Success) {
     assert(_recursions == 0, "invariant");
     return true;
   }
 
-  if (r == TryLockResult::HasOwner && owner() == current) {
-    _recursions++;
-    return true;
-  }
+  // Set check_owner to false (it's default value is true) if you want
+  // to use ObjectMonitor::try_enter() as a public way of doing TryLock().
+  // Used this way in SharedRuntime::monitor_exit_helper().
+  if (check_owner) {
```

Comment:
Probably preference but an early return here is easier for me to parse.

Reply:
I agree.